src/Sarif.Multitool/Rules/SARIF1004.ExpressUriBaseIdsCorrectly.cs

```diff
@@ -63,14 +70,48 @@ private void AnalyzeOriginalUriBaseIdsEntry(string uriBaseId, ArtifactLocation a
             {
                 var uri = new Uri(uriString, UriKind.RelativeOrAbsolute);
 
+                // TopLevelUriBaseIdMustBeAbsolute: Top level uriBaseId must be absolute.
                 if (artifactLocation.UriBaseId == null && !uri.IsAbsoluteUri)
                 {
-                    LogResult(pointer, nameof(RuleResources.SARIF1004_ExpressUriBaseIdsCorrectly_Error_TopLevelUriBaseIdMustBeAbsolute_Text), uriString, uriBaseId);
+                    // {0}: The URI '{1}' belonging to the '{2}' element of run.originalUriBaseIds is not an absolute URI.
+                    LogResult(
+                        pointer,
+                        nameof(RuleResources.SARIF1004_ExpressUriBaseIdsCorrectly_Error_TopLevelUriBaseIdMustBeAbsolute_Text),
+                        uriString,
+                        uriBaseId);
                 }
 
+                // UriBaseIdValueMustEndWithSlash: uriBaseIds must end with a slash.
                 if (!uriString.EndsWith("/"))
                 {
-                    LogResult(pointer, nameof(RuleResources.SARIF1004_ExpressUriBaseIdsCorrectly_Error_UriBaseIdValueMustEndWithSlash_Text), uriString, uriBaseId);
+                    // {0}: The URI '{1}' belonging to the '{2}' element of run.originalUriBaseIds does not end with a slash.
+                    LogResult(
+                        pointer,
+                        nameof(RuleResources.SARIF1004_ExpressUriBaseIdsCorrectly_Error_UriBaseIdValueMustEndWithSlash_Text),
+                        uriString,
+                        uriBaseId);
+                }
+
+                // UriBaseIdValueMustNotContainDotDotSegment: uriBaseIds must not contain `..` segment(s).
+                if (uriString.Contains(".."))
```

Comment:
>Contains [](start = 30, length = 8)

I suggest a stricter test: split on '/' and look for segments that are _exactly_ `'..'`.

`Uri` object actually has a `Segments` property but sadly it only works for absolute URIs.